libcudacxx/include/cuda/std/__algorithm/sort.h

```diff
@@ -0,0 +1,1129 @@
+template <class _Compare, class _RandomAccessIterator>
+_LIBCUDACXX_HIDE_FROM_ABI void
+__partially_sorted_swap(_RandomAccessIterator __x, _RandomAccessIterator __y, _RandomAccessIterator __z, _Compare __c)
+{
+  // Note: this function behaves correctly even with proxy iterators (because it relies on `value_type`).
+  using value_type = typename iterator_traits<_RandomAccessIterator>::value_type;
+  bool __r         = __c(*__z, *__x);
+  value_type __tmp = __r ? *__z : *__x;
+  *__z             = __r ? *__x : *__z;
+  __r              = __c(__tmp, *__y);
+  *__x             = __r ? *__x : *__y;
+  *__y             = __r ? *__y : __tmp;
+}
+
+template <class,
+          class _Compare,
+          class _RandomAccessIterator,
+          enable_if_t<__use_branchless_sort<_Compare, _RandomAccessIterator>::value, int> = 0>
+_LIBCUDACXX_HIDE_FROM_ABI void __sort3_maybe_branchless(
+  _RandomAccessIterator __x1, _RandomAccessIterator __x2, _RandomAccessIterator __x3, _Compare __c)
+{
+  _CUDA_VSTD::__cond_swap<_Compare>(__x2, __x3, __c);
+  _CUDA_VSTD::__partially_sorted_swap<_Compare>(__x1, __x2, __x3, __c);
+}
+
+template <class _AlgPolicy,
+          class _Compare,
+          class _RandomAccessIterator,
+          enable_if_t<!__use_branchless_sort<_Compare, _RandomAccessIterator>::value, int> = 0>
+_LIBCUDACXX_HIDE_FROM_ABI void __sort3_maybe_branchless(
+  _RandomAccessIterator __x1, _RandomAccessIterator __x2, _RandomAccessIterator __x3, _Compare __c)
+{
+  _CUDA_VSTD::__sort3<_AlgPolicy, _Compare>(__x1, __x2, __x3, __c);
+}
+
+template <class,
+          class _Compare,
+          class _RandomAccessIterator,
+          enable_if_t<__use_branchless_sort<_Compare, _RandomAccessIterator>::value, int> = 0>
+_LIBCUDACXX_HIDE_FROM_ABI void __sort4_maybe_branchless(
+  _RandomAccessIterator __x1,
+  _RandomAccessIterator __x2,
+  _RandomAccessIterator __x3,
+  _RandomAccessIterator __x4,
+  _Compare __c)
+{
+  _CUDA_VSTD::__cond_swap<_Compare>(__x1, __x3, __c);
+  _CUDA_VSTD::__cond_swap<_Compare>(__x2, __x4, __c);
+  _CUDA_VSTD::__cond_swap<_Compare>(__x1, __x2, __c);
+  _CUDA_VSTD::__cond_swap<_Compare>(__x3, __x4, __c);
+  _CUDA_VSTD::__cond_swap<_Compare>(__x2, __x3, __c);
+}
+
+template <class _AlgPolicy,
+          class _Compare,
+          class _RandomAccessIterator,
+          enable_if_t<!__use_branchless_sort<_Compare, _RandomAccessIterator>::value, int> = 0>
+_LIBCUDACXX_HIDE_FROM_ABI void __sort4_maybe_branchless(
+  _RandomAccessIterator __x1,
+  _RandomAccessIterator __x2,
+  _RandomAccessIterator __x3,
+  _RandomAccessIterator __x4,
+  _Compare __c)
+{
+  _CUDA_VSTD::__sort4<_AlgPolicy, _Compare>(__x1, __x2, __x3, __x4, __c);
+}
+
+template <class _AlgPolicy,
+          class _Compare,
+          class _RandomAccessIterator,
+          enable_if_t<__use_branchless_sort<_Compare, _RandomAccessIterator>::value, int> = 0>
+_LIBCUDACXX_HIDE_FROM_ABI void __sort5_maybe_branchless(
```

Comment:
Q: Should we merge this function with the following one and use `if constexpr(__use_branchless_sort<....>)`. Probably applies to other places as well?